enable mkldnn conv2d backward to support mkldnn tensor input

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48994 enable mkldnn conv2d backward to support mkldnn tensor input**
* #48922 BFloat16: enable prepacked weights's inference

Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

XiaobingSuper committed:
enable mkldnn conv2d backward to support mkldnn tensor input

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -99,7 +99,8 @@ struct TORCH_API AccumulateGrad : public Node {
           new_grad.use_count() <= num_expected_refs &&
-          utils::obeys_layout_contract(new_grad, variable)) {
+          ((!new_grad.is_mkldnn() &&utils::obeys_layout_contract(new_grad, variable))
+           || new_grad.is_mkldnn())){
```
Does this mean that the layout contract is not respected for mkldnn Tensors? Should we update the comment describing it above to reflect that?

XiaobingSuper replied:
Yes, MKLDNN tensor is a opaque tensor which is not obey the layout contract.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -130,7 +131,7 @@ struct TORCH_API AccumulateGrad : public Node {
             new_grad.options()));
       } else {
-        if (new_grad.is_sparse()) {
+        if (new_grad.is_sparse() || new_grad.is_mkldnn()) {
```
Why is mkldnn Tensors associated with dense Tensors in the conditions above but to sparse Tensors here?

XiaobingSuper replied:
**clone_obey_contract** in dense tensor step seems will check tensor's layout which will clone a tensor having same layout wth variable , but MKLDNN tensor has not layout attribute, so associated with sparse tensor here.

albanD replied:
Would it be clearer to separate these? Also updating the doc above would make these choices clearer.

albanD commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -136,15 +138,24 @@ Tensor mkldnn_convolution_backward_input(
     IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups, bool bias_defined)
 {
+  // for training case, grad_output and input can be cpu tensor or MKLDNN tensor,
+  // but weight and bias always cpu tensor.
```
nit: missing a verb

jgong5 commented:
@VitalyFedyunin

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

ngimel commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -100,11 +100,13 @@ struct TORCH_API AccumulateGrad : public Node {
           !new_grad.is_sparse() &&
           new_grad.use_count() <= num_expected_refs &&
-          utils::obeys_layout_contract(new_grad, variable)) {
+          ((!new_grad.is_mkldnn() && utils::obeys_layout_contract(new_grad, variable))
```
HOw can grad be an mkldnn tensor, when mkldnn_convolution_backward_weights always converts grad to dense tensor? https://github.com/pytorch/pytorch/blob/b7449f9a21cc799b7f5632db26b74991991668df/aten/src/ATen/native/mkldnn/Conv.cpp#L193

XiaobingSuper replied:
hi, @ngimel, Could you give me more details about it? I can't catch your meaning.

ngimel replied:
This function (`accumulateGrad`) gets grads returned backward functions (in case of mkl convolutions that would be `mkl_convolution_backward_weights`). Since `mkl_convolution_backward_weights` always returns dense tensors, `new_grad` cannot be mkldnn tensor, unless I'm missing something?

XiaobingSuper replied:
@ngimel, yes, **mkldnn_convolution_backward_weight**s always returns dense tensors, but for **mkldnn_convolution_backward_input**,  it returns mkldnn tensor.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

ngimel commented on test/test_mkldnn.py:
```diff
@@ -176,14 +176,36 @@ def _test_conv_base(self, dim):
+            if not train:
+                self._test_serialization(mkldnn_conv, (x.to_mkldnn(),))
+                self._test_tracing(mkldnn_conv, (x.to_mkldnn(),))
+            elif train and dim != 1:
```
you don't need `train` conditional here

XiaobingSuper replied:
Changed.

ngimel commented on test/test_mkldnn.py:
```diff
@@ -176,14 +176,36 @@ def _test_conv_base(self, dim):
+            elif train and dim != 1:
+                loss2 = y_mkldnn.sum()
+                loss2.backward()
+                self.assertEqual(x1.grad, x2.grad.to_dense())
```
please also add test that x2.grad is an mkldnn tensor.

XiaobingSuper replied:
Added.

ngimel commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -100,11 +100,13 @@ struct TORCH_API AccumulateGrad : public Node {
           !new_grad.is_sparse() &&
           new_grad.use_count() <= num_expected_refs &&
-          utils::obeys_layout_contract(new_grad, variable)) {
+          ((!new_grad.is_mkldnn() && utils::obeys_layout_contract(new_grad, variable))
```
you can write this conditional as 
```
new_grad.is_mkldnn() || utils::obeys_layout_contract(new_grad, variable)
```

XiaobingSuper replied:
Thanks, I changed it.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -166,7 +172,7 @@ struct TORCH_API AccumulateGrad : public Node {
         // 3. `variable_grad` is dense, and `new_grad` is dense.
         //
-        // In all of these three cases, `variable_grad += new_grad` is a
+        // In all of these four cases, `variable_grad += new_grad` is a
```
yes, I add mkldnn tensor case now. Thanks.

VitalyFedyunin commented:
Please rebase entire stack, I'm getting internal merge conflicts. thanks.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper commented:
> Please rebase entire stack, I'm getting internal merge conflicts. thanks.

Rebased, thanks.

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "enable mkldnn conv2d backward to support mkldnn tensor input"


Differential Revision: [D25537189](https://our.internmc.facebook.com/intern/diff/D25537189)

[ghstack-poisoned]
